### DISCUSSION TITLE
SSO (SAML or OIDC) with Cockpit?

### DISCUSSION THREAD

**Discussion by ygini** [cockpit-project/cockpit#19056] [Q&A]
Hello,

I'm interested to deploy cockpit to help our production team to easily manage edge nodes in a decentralized context. 

We don't have legacy directory service like AD or anything using Kerberos. 

Also, we just don't have passwords. All our authentication are using so-called "modern" authentication, which is basically federation via SAML or OIDC to our IDP. 

And then the IDP will use mTLS to authenticate the user and will handle role based rules to allow or not access to a service. 

I would like to integrate Cockpit to that context: having website for cockpit doing SSO and allowing our team to access all host capabilities without the need to know a password. 

I see two option to do so:
- native OIDC / SAML support in cockpit or via a plug-in
- reverse proxy with some HTTP Header to pass the user ID

Is there a chance to have one or the other mechanism working? Or something else?

Cheers

**Comment by martinpitt** [comment]
So far I don't know how SAML or OIDC work. But as you mentioned mTLS, cockpit does support [TLS certificate based authentication](https://cockpit-project.org/guide/latest/cert-authentication.html) (like with smartcards). That uses the "reverse proxy with some HTTP Header to pass the user ID" approach, [cockpit-tls](https://cockpit-project.org/guide/latest/cockpit-tls.8.html) implements that.

Other than that, you can also implement any custom authentication schema and configure it in [cockpit.conf](https://github.com/cockpit-project/cockpit/blob/main/doc/authentication.md). This supports e.g. Bearer token (for SSO or oauth2). However, this is not well documented nor are there a lot of good existing examples. Sorry for lack of details, just wanted to paint some initial broad strokes in which direction this could go.

**Reply by ygini** [reply]
mTLS implementation inside cockpit is not appropriate for many reason (especially the management of the user database).

And yes, I've seen "we can implement our own", but I'm just disapointed to see no default implementation of this "own" for largely used standard like SAML or OIDC…

**Reply by jelly** [reply]
Cockpit was developed to fully integrate with PAM as auth mechanism, so that is another route you could take using `pam_oath`. I have no experience with that PAM module however.

**Reply by mwmahlberg** [reply]
@jelly Stating the obvious sometimes helps to remind one of the basics.

**Comment by SEJeff** [comment]
It would be super helpful to auth cockpit through [google workspace SSO](https://support.google.com/a/answer/12032922?hl=en) flow using SAML or OIDC as @ygini suggested.

**Comment by vaminakov** [comment]
Cockpit using standart linux authentication, so all needed - deploy oidc via pam. I'm searching right way for it now.

**Reply by moutasem1989** [reply]
have you managed to get it working?

**Reply by vaminakov** [reply]
Not yet

**Comment by cxfcxf** [comment]
> Cockpit using standart linux authentication, so all needed - deploy oidc via pam. I'm searching right way for it now.

so i don't think that's accurate, pam_oidc implants login through oidc, but it doesn't implant SSO

in each server you can login through pam_odic when you provide a token, but you will have to provide that token every single time for each server session on each server.

same as you type password, you input it every single time.

**Comment by moutasem1989** [comment]
In NPM, I added this for services that do not support OIDC but has basic HTTP authentication. I then integrated it into Authentik reverse proxy authentication and it works as expected:

```
location /login {
    proxy_pass  $forward_scheme://$server:$port/login;
    proxy_set_header Authorization "Basic wersdfxcvetc";     #base64 of username:password
    proxy_pass_header Authorization;
}
```

Perhaps it is feasible to use a similar approach with a reverse proxy authentication ? Is there a way to pass headers with correct values to bypass authentication?

check [this](https://github.com/cockpit-project/cockpit/blob/main/doc/login.md).

**Reply by moutasem1989** [reply]
this is the working custom configuration of NPM with Authentik reverse proxy authentication 

```
# Increase buffer size for large headers
# This is needed only if you get 'upstream sent too big header while reading response
# header from upstream' error when trying to access an application protected by goauthentik
proxy_buffers 8 16k;
proxy_buffer_size 32k;

# Make sure not to redirect traffic to a port 4443
port_in_redirect off;

location / {
    # Put your proxy_pass to your application here
    # proxy_pass          $forward_scheme://$server:$port;
    # Set any other headers your application might need
    # proxy_set_header Host $host;
    # proxy_set_header ...
    proxy_set_header        Host $host;
    proxy_set_header Authorization "Basic abcxyz==";  #base64 of user:password 
    proxy_pass_header Authorization;
    proxy_pass  $forward_scheme://$server:$port;

    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;

    ##############################
    # authentik-specific config
    ##############################
    auth_request     /outpost.goauthentik.io/auth/nginx;
    error_page       401 = @goauthentik_proxy_signin;
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header       Set-Cookie $auth_cookie;

    # translate headers from the outposts back to the actual upstream
    auth_request_set $authentik_username $upstream_http_x_authentik_username;
    auth_request_set $authentik_groups $upstream_http_x_authentik_groups;
    auth_request_set $authentik_email $upstream_http_x_authentik_email;
    auth_request_set $authentik_name $upstream_http_x_authentik_name;
    auth_request_set $authentik_uid $upstream_http_x_authentik_uid;

    proxy_set_header X-authentik-username $authentik_username;
    proxy_set_header X-authentik-groups $authentik_groups;
    proxy_set_header X-authentik-email $authentik_email;
    proxy_set_header X-authentik-name $authentik_name;
    proxy_set_header X-authentik-uid $authentik_uid;
    #add_header X-calibre-web-user "CloudAdmin" always;
}

# all requests to /outpost.goauthentik.io must be accessible without authentication
location /outpost.goauthentik.io {
    proxy_pass              https://authentik-server:9443/outpost.goauthentik.io;
    # ensure the host of this vserver matches your external URL you've configured
    # in authentik
    proxy_set_header        Host $host;
    proxy_set_header        X-Original-URL $scheme://$http_host$request_uri;
    add_header              Set-Cookie $auth_cookie;
    auth_request_set        $auth_cookie $upstream_http_set_cookie;
    proxy_pass_request_body off;
    proxy_set_header        Content-Length "";
}

# Special location for when the /auth endpoint returns a 401,
# redirect to the /start URL which initiates SSO
location @goauthentik_proxy_signin {
    internal;
    add_header Set-Cookie $auth_cookie;
    return 302 /outpost.goauthentik.io/start?rd=$request_uri;
    # For domain level, use the below error_page to redirect to your authentik server with the full redirect path
    # return 302 https://authentik.company/outpost.goauthentik.io/start?rd=$scheme://$http_host$request_uri;
}
```
it auto authenticates in to Cockpit and then do a reverse proxy authentication with Authentik

**Reply by GerardWoldendorp** [reply]
@moutasem1989 Hi, how did you then set it up eventually, in NPM and authentik?

**Comment by moutasem1989** [comment]
@GerardWoldendorp go to Group or User. Click on edit and scroll down to Attributes and add the following:

```
cockpit_password: cockpituserpassword
cockpit_username: cockpitusername
```
Then created a Proxy authentication provider in Authentik. Under authentication add the attributes we just created: `cockpit_password` and `cockpit_username`. There are pleats of documentation on how to create proxy authentication providers.

For NPM I added as follows:

```
# Increase buffer size for large headers
# This is needed only if you get 'upstream sent too big header while reading response
# header from upstream' error when trying to access an application protected by goauthentik
proxy_buffers 8 16k;
proxy_buffer_size 32k;

proxy_connect_timeout 600;
proxy_send_timeout 600;
proxy_read_timeout 600;
send_timeout 600;

# Make sure not to redirect traffic to a port 4443
port_in_redirect off;

location / {
    proxy_pass  $forward_scheme://$server:$port;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;

    ##############################
    # authentik-specific config
    ##############################
    auth_request     /outpost.goauthentik.io/auth/nginx;
    error_page       401 = @goauthentik_proxy_signin;
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header       Set-Cookie $auth_cookie;
    auth_request_set $authentik_auth $upstream_http_authorization;
    proxy_set_header Authorization $authentik_auth;
    proxy_pass_header Authorization;
}

# all requests to /outpost.goauthentik.io must be accessible without authentication
location /outpost.goauthentik.io {
    proxy_pass              https://authentik-server:9443/outpost.goauthentik.io;
    # ensure the host of this vserver matches your external URL you've configured
    # in authentik
    proxy_set_header        Host $host;
    proxy_set_header        X-Original-URL $scheme://$http_host$request_uri;
    add_header              Set-Cookie $auth_cookie;
    auth_request_set        $auth_cookie $upstream_http_set_cookie;
    proxy_pass_request_body off;
    proxy_set_header        Content-Length "";
}

# Special location for when the /auth endpoint returns a 401,
# redirect to the /start URL which initiates SSO
location @goauthentik_proxy_signin {
    internal;
    add_header Set-Cookie $auth_cookie;
    return 302 /outpost.goauthentik.io/start?rd=$request_uri;
    # For domain level, use the below error_page to redirect to your authentik server with the full redirect path
    # return 302 https://authentik.company/outpost.goauthentik.io/start?rd=$scheme://$http_host$request_uri;
}
```

try clearing cookies or try in incognito mode if things don’t work the first time.

**Reply by begunfx** [reply]
Finally got this to work.  I did use the specific headers that you included, but did not use the increased timeouts and it worked like a charm.  Thanks!
